File: IPaImageURLLoader/IPaImageURLLoader/IPaImageURLButton.m
//
//  IPaImageURLButton.m
//  IPaImageURLLoader
//
//  Created by IPa Chen on 2014/10/12.
//  Copyright (c) 2014年 IPaPa. All rights reserved.
//

#import "IPaImageURLButton.h"
#import "IPaImageURLLoader.h"
@implementation IPaImageURLButton
{
    id imageObserver;
}
-(void)awakeFromNib
{
    [super awakeFromNib];
    __weak IPaImageURLButton *weakSelf = self;
    imageObserver = [[NSNotificationCenter defaultCenter] addObserverForName:IPA_NOTIFICATION_IMAGE_LOADED object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification* noti){
        NSString *imageID = noti.userInfo[IPA_NOTIFICATION_KEY_IMAGEID];
        if ([imageID isEqualToString:weakSelf.imageURL]) {
            
            [weakSelf setImage:noti.userInfo[IPA_NOTIFICATION_KEY_IMAGE] forState:UIControlStateNormal];
        }
        else if ([imageID isEqualToString:weakSelf.backgroundImageURL]) {
            [weakSelf setBackgroundImage:noti.userInfo[IPA_NOTIFICATION_KEY_IMAGE] forState:UIControlStateNormal];
        }
    }];
}
-(void)setImageURL:(NSString *)imageURL
{
    _imageURL = imageURL;
    [self setImage:[[IPaImageURLLoader defaultLoader] loadImageFromURL:self.imageURL] forState:UIControlStateNormal];
    
}
-(void)setBackgroundImageURL:(NSString *)backgroundImageURL
{
    _backgroundImageURL = backgroundImageURL;
    [self setBackgroundImage:[[IPaImageURLLoader defaultLoader] loadImageFromURL:self.imageURL] forState:UIControlStateNormal];
    
}
-(void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:imageObserver];
}
@end

File: IPaImageURLLoader/IPaImageURLLoader/IPaImageURLButton.swift
//
//  IPaImageURLButton.swift
//  IPaImageURLLoader
//
//  Created by IPa Chen on 2015/6/16.
//  Copyright (c) 2015年 AMagicStudio. All rights reserved.
//

import Foundation
import UIKit
public class IPaImageURLButton : UIButton {
    private var _imageURL:String?
    private var _backgroundImageURL:String?
    private var imageObserver:NSObjectProtocol?
    public var imageURL:String? {
        get {
            return _imageURL
        }
        set {
            setImageURL(newValue, defaultImage: nil)
        }
    }
    public var backgroundImageURL:String? {
        get {
            return _backgroundImageURL
        }
        set {
            setBackgroundImageURL(newValue, defaultImage: nil)
        }
    }
    deinit {
        if let imageObserver = imageObserver {
            NotificationCenter .default
                .removeObserver(imageObserver)
        }
    }
    func createImageObserver () {
        if imageObserver != nil {
            return
        }
        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: IPA_NOTIFICATION_IMAGE_LOADED), object: nil, queue: OperationQueue.main, using: {
            noti in
            if let userInfo = noti.userInfo {
                let imageID = userInfo[IPA_NOTIFICATION_KEY_IMAGEID] as! String
                
                if let imageURL = self.imageURL {
                    if imageID == imageURL {
                        do {
                            let data = try Data(contentsOf: (userInfo[IPA_NOTIFICATION_KEY_IMAGEFILEURL] as! URL))
                            let image = UIImage(data: data)
                            self.setImage(image, for: .normal)
                            
                        }

                        catch {
                            
                        }
                            
                    
                    }
                }
                if let backgroundImageURL = self.backgroundImageURL {
                    if imageID == backgroundImageURL {
                        do {
                            let data = try Data(contentsOf: (userInfo[IPA_NOTIFICATION_KEY_IMAGEFILEURL] as! URL))
                            let image = UIImage(data: data)
                            self.setBackgroundImage(image, for: .normal)
                        }
                        catch {
                            
                        }
                    }
                }
            }
        })
    }
    public func setImageURL(_ imageURL:String?,defaultImage:UIImage?) {
        createImageObserver()
        _imageURL = imageURL
        var image:UIImage?
        if let imageURL = imageURL {
            image = IPaImageURLLoader.sharedInstance.loadImage(url: (imageURL as NSString).addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed)!, imageID: imageURL)
            
        }
        
        setImage((image == nil) ? defaultImage :image, for: .normal)
    }
    public func setBackgroundImageURL(_ imageURL:String?,defaultImage:UIImage?) {
        createImageObserver()
        _backgroundImageURL = imageURL
        var image:UIImage?
        if let imageURL = imageURL {
            image = IPaImageURLLoader.sharedInstance.loadImage(url: (imageURL as NSString).addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed)!, imageID: imageURL)
            
        }
        
        setBackgroundImage((image == nil) ? defaultImage :image, for: .normal)
    }
}

